RHOAIENG-11155: Better explanation of 'Authorize Access' UI

daniellutz opened the pull request:
This feature will improve the user experience in a way that the user required OAuth scope will change from a UI showing the scopes to a simple login confirmation page, according to [https://issues.redhat.com/browse/RHOAIENG-11155](https://issues.redhat.com/browse/RHOAIENG-11155)

## Description
There is an option to inject the OAuth scope into the proxy sidecar container, in a way that it will be required only for the user to confirm his login to accept it, instead of showing up a page with confusing permissions and a bad user experience.

Not only the OAuth scope need to be passed on, but also a volume need to be mounted to gather the OAuth client secret, in a way that the application understands who is authenticating properly.

## How Has This Been Tested?
Manual tests have been executed, using the following `devFlags`, with a clean test running in OpenShift Local environment with RHOAI 2.16.0 installed:

```yaml
    workbenches:
      managementState: Managed
      devFlags:
        manifests:
          - contextDir: components/odh-notebook-controller/config
            sourcePath: ''
            uri: 'https://github.com/daniellutz/odh-kubeflow/archive/authorize-access-ui.tar.gz'
```

## Merge criteria:
- [X] The commits are squashed in a cohesive manner and have meaningful messages.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has manually tested the changes and verified that the changes work

atheo89 commented:
Thank for opening this PR! 
Seems that in my end there is no `oauth-client` data in order to be mounted and start properly the oauth proxy. 

Steps to reproduce:
1. Replace odh-notebook-controller image with the one that generated from this PR.
2. Create a new notebook
3. Then I got:  
![image](https://github.com/user-attachments/assets/08830506-190f-4bea-b427-5eeb2b63e749)

4. Logs from oauth conteiner show: 
```
2024/11/14 08:31:30 provider.go:120: Defaulting client-id to system:serviceaccount:test:q2
2024/11/14 08:31:30 provider.go:125: Defaulting client-secret to service account token /var/run/secrets/kubernetes.io/serviceaccount/token
2024/11/14 08:31:30 main.go:140: Invalid configuration:
  cannot read client-secret-file: open /etc/oauth/client/secret: no such file or directory
```
5. 
```
atheodor@fedora:~-$ oc get secrets
NAME                       TYPE                      DATA   AGE
builder-dockercfg-jkztm    kubernetes.io/dockercfg   1      7d20h
default-dockercfg-knsk6    kubernetes.io/dockercfg   1      7d20h
deployer-dockercfg-9ngd6   kubernetes.io/dockercfg   1      7d20h
pipeline-dockercfg-5fn4m   kubernetes.io/dockercfg   1      7d20h
q2-dockercfg-2x9gd         kubernetes.io/dockercfg   1      35m
q2-oauth-client            Opaque                    0      35m      <---- Problematic secret
q2-oauth-config            Opaque                    1      35m
q2-tls                     kubernetes.io/tls         2      35m

```
6.  Oauth-client didn't populated data secret and thats why failed

What i missed?

jiridanek commented on components/odh-notebook-controller/config/base/params.env:
```diff
@@ -1 +1 @@
-odh-notebook-controller-image=quay.io/opendatahub/odh-notebook-controller:main-3f931d2
+odh-notebook-controller-image=quay.io/dlutz/odh-notebook-controller:authorize-access
```
testing purposes? where? what?

jiridanek commented on components/odh-notebook-controller/controllers/notebook_controller_test.go:
```diff
@@ -646,6 +646,15 @@ var _ = Describe("The Openshift Notebook controller", func() {
+								VolumeSource: corev1.VolumeSource{
+									Secret: &corev1.SecretVolumeSource{
+										SecretName:  Name + "-oauth-client-generated",
+										DefaultMode: pointer.Int32Ptr(420),
```
@daniellutz the idea here is that the linter only runs on new code, so that's why it is not flagging the existing usages of `pointer.Int32Ptr`. What you should do here is to simply use `ptr.To` instead of this, and it will work just fine. The idea is that ptr.To is a better replacement for the old functions, and it could not be used before because it requires go 1.18+ features https://pkg.go.dev/k8s.io/utils/ptr#section-readme

jiridanek commented:
@daniellutz regarding ODH Notebook Controller Integration Test / build (pull_request) Failing after 14m

That's never going to pass because in the test we are running in a KinD cluster (https://kind.sigs.k8s.io/) and we only have the notebook controller and no other components of rhoai, most importantly we don't have rhods-operator that would create your oauth secret for the pod; so the solution should be to create the secret in test setup. I'll take a look r.n.

atheo89 commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -209,6 +209,26 @@ func NewNotebookOAuthSecret(notebook *nbv1.Notebook) *corev1.Secret {
+				"notebook-name": notebook.Name,
+			},
+			Annotations: map[string]string{
+				"secret-generator.opendatahub.io/name":               "secret",
```
FYI: This specific annotations used by rhoai-operator to create a secret `secret-generated` 

https://github.com/opendatahub-io/opendatahub-operator/tree/84d22f35a620f43f0e8b397b1b45e2bdb25a8f46/controllers/secretgenerator#basic-usage

harshad16 commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -209,6 +209,26 @@ func NewNotebookOAuthSecret(notebook *nbv1.Notebook) *corev1.Secret {
 }
 
+// NewNotebookOAuthClientSecret defines the desired OAuth client secret object
+func NewNotebookOAuthClientSecret(notebook *nbv1.Notebook) *corev1.Secret {
```
As we would not be utilizing, opendatahub-operator secretgenerator, 
lets change the logic here, and write our own secret create

please take this as reference: https://github.com/opendatahub-io/kubeflow/blob/30afc1574e05bb05b21565db64e3b0ba5d6b9060/components/odh-notebook-controller/controllers/notebook_oauth.go#L190

and adjust this function , and for secret generation, 
utilize the logic of random function from here: https://github.com/opendatahub-io/opendatahub-operator/blob/c1671ab5fd11baea814f8acdee1bc448d502fb1c/controllers/secretgenerator/secret.go#L91
```suggestion
func NewNotebookOAuthClientSecret(notebook *nbv1.Notebook) *corev1.Secret {
  // Generate the client secret for the OAuth proxy
	randomValue := make([]byte, 32)
		for i := 0; i < secret.Complexity; i++ {
			num, err := rand.Int(rand.Reader, big.NewInt(int64(len(letterRunes))))
			if err != nil {
				return err
			}
			randomValue[i] = letterRunes[num.Int64()]
		}
		
	// Create a Kubernetes secret to store the cookie secret
	return &corev1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name:      notebook.Name + "-oauth-client",
			Namespace: notebook.Namespace,
			Labels: map[string]string{
				"notebook-name": notebook.Name,
			},
		},
		StringData: map[string]string{
			"secret": string(randomValue),
		},
	}
```

and adjust the oauth-proxy to directly pick value from this secret

harshad16 reviewed (changes requested):
Based on the discussion with platform, team 
we got to know, it would better if the logic of secret is on own side, so lets adjust that

daniellutz committed:
Add methods to handle OAuth client secret generation

- Add volumes to store the oauth-client configuration;
- Add extra parameters to container creation, including --client-id, --client-secret and --scope;
- Add method to generate the secret randomly when authenticating using OAuth;
- Add link between the notebook's route, user's generated oauth secret and OAuthClient config;

daniellutz commented:
I've had to change the methods from what we were trying before, due to unplanned changes from other teams to namespace watching to the secret generation work.

Now, the script will get the notebook's route, automatically generate the secret, create the OAuthClient link between the route and the secret and enable the access without requesting permissions with the UI.

Reach out in any case of questions, suggestions, etc.

jiridanek commented:
Looks like the test fails are legit

```
--- FAIL: TestE2ENotebookController (394.17s)
    --- PASS: TestE2ENotebookController/validate_controllers (20.45s)
        --- PASS: TestE2ENotebookController/validate_controllers/Validate_Kubeflow_notebook_controller (10.40s)
        --- PASS: TestE2ENotebookController/validate_controllers/Validate_ODH_notebook_controller (10.04s)
    --- FAIL: TestE2ENotebookController/create (282.30s)
        --- FAIL: TestE2ENotebookController/create/thoth-minimal-oauth-notebook (282.30s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Creation_of_Notebook_instance (10.26s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_Route_Validation (10.08s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_Network_Policies_Validation (20.08s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_Statefulset_Validation (10.05s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_OAuth_sidecar_Validation (0.04s)
            --- FAIL: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Verify_Notebook_Traffic (10.67s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Verify_Notebook_Culling (221.13s)
    --- FAIL: TestE2ENotebookController/update (61.08s)
        --- FAIL: TestE2ENotebookController/update/thoth-minimal-oauth-notebook (61.07s)
            --- PASS: TestE2ENotebookController/update/thoth-minimal-oauth-notebook/Update_Notebook_instance (10.18s)
            --- PASS: TestE2ENotebookController/update/thoth-minimal-oauth-notebook/Notebook_Route_Validation_After_Update (10.08s)
            --- PASS: TestE2ENotebookController/update/thoth-minimal-oauth-notebook/Notebook_Network_Policies_Validation_After_Update (20.08s)
            --- PASS: TestE2ENotebookController/update/thoth-minimal-oauth-notebook/Notebook_Statefulset_Validation_After_Update (10.04s)
            --- PASS: TestE2ENotebookController/update/thoth-minimal-oauth-notebook/Notebook_OAuth_sidecar_Validation_After_Update (0.04s)
            --- FAIL: TestE2ENotebookController/update/thoth-minimal-oauth-notebook/Verify_Notebook_Traffic_After_Update (10.64s)
    --- PASS: TestE2ENotebookController/delete (30.34s)
        --- PASS: TestE2ENotebookController/delete/thoth-minimal-oauth-notebook (30.34s)
            --- PASS: TestE2ENotebookController/delete/thoth-minimal-oauth-notebook/Notebook_Deletion (0.13s)
            --- PASS: TestE2ENotebookController/delete/thoth-minimal-oauth-notebook/Dependent_Resource_Deletion (30.21s)
FAIL
FAIL	github.com/opendatahub-io/kubeflow/components/odh-notebook-controller/e2e	394.196s
```

https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/opendatahub-io_kubeflow/449/pull-ci-opendatahub-io-kubeflow-main-odh-notebook-controller-e2e/1878640725080412160

This is running only on openshift-ci, that's why gha is passing.

jiridanek commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -209,39 +227,92 @@ func NewNotebookOAuthSecret(notebook *nbv1.Notebook) *corev1.Secret {
 
-	// Create the OAuth secret if it does not already exist
-	foundSecret := &corev1.Secret{}
+func (r *OpenshiftNotebookReconciler) createSecret(notebook *nbv1.Notebook, ctx context.Context, desiredSecret *corev1.Secret) error {
```
```suggestion
func (r *OpenshiftNotebookReconciler) reconcileSecret(notebook *nbv1.Notebook, ctx context.Context, desiredSecret *corev1.Secret) error {
```

The function creates secret _only if_ it does not already exist. So the name should reflect that.

daniellutz replied:
I agree, it's a better name, I will refactor this method name. Thanks!

jiridanek commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -264,3 +335,61 @@ func (r *OpenshiftNotebookReconciler) ReconcileOAuthRoute(
+func (r *OpenshiftNotebookReconciler) createOAuthClient(notebook *nbv1.Notebook, ctx context.Context) error {
+	log := logf.FromContext(ctx)
+
+	//
```
do these empty comment lines mean anything? just asking

daniellutz replied:
Removing them. Thanks.

jiridanek reviewed (changes requested):
lgtm, the only thing that really should be fixed is the ocp-ci e2e test, but other than that it's good to go imo

jiridanek commented:
I'm trying the PR images

This is what happens with running notebook if I switch from regular controllers in 2.17 nightly to the pr ones

```
{"level":"info","ts":"2025-01-14T09:17:30Z","logger":"controllers.Notebook","msg":"Update blocked, notebook pod template would be changed by the webhook","notebook":"mywb2","namespace":"vscodedocs","diff":"{v1.PodSpec}.Volumes[5->?]: {Name:oauth-client VolumeSource:{HostPath:nil EmptyDir:nil GCEPersistentDisk:nil AWSElasticBlockStore:nil GitRepo:nil Secret:&SecretVolumeSource{SecretName:mywb2-oauth-client,Items:[]KeyToPath{},DefaultMode:*420,Optional:nil,} NFS:nil ISCSI:nil Glusterfs:nil PersistentVolumeClaim:nil RBD:nil FlexVolume:nil Cinder:nil CephFS:nil Flocker:nil DownwardAPI:nil FC:nil AzureFile:nil ConfigMap:nil VsphereVolume:nil Quobyte:nil AzureDisk:nil PhotonPersistentDisk:nil Projected:nil PortworxVolume:nil ScaleIO:nil StorageOS:nil CSI:nil Ephemeral:nil}} != <invalid reflect.Value>"}
```

not very understandable to read, but I guess it's refusing to mount the oauth config as a volume, which is correct, to prevent pod restart. And pod was not restarted, so that's good.

One more thing, what happens if i start the pod with old controller, then switch to new controller images, and only then I try to access the workbench (with all the oauth flow)?

And that did immediately start working for me, no oauth dialog. Even without restarting the workbench container, apparently, the oauth prompt did not appear! I immediately got thrown into Jupyter. I'm shocked, will try this again.

Starting and opening a brand new workbench gave me

```
{"error":"unauthorized_client","error_description":"The client is not authorized to request a token using this method.","state":"8539f027f67f47bfd5efbb1ebd30259f:/notebook/vscodedocs/noreg1"}
```

FAIL!

daniellutz commented:
As soon as possible I will check this scenario and try to fix it. Thank you very much for this test, Jiri.

jiridanek commented:
I'm going to use

```
    workbenches:
      devFlags:
        manifests:
          - contextDir: components/odh-notebook-controller/config
            sourcePath: ''
            uri: 'https://github.com/daniellutz/odh-kubeflow/archive/authorize-access-ui.tar.gz'
```

and change the odh notebook controller manually to

* quay.io/opendatahub/odh-notebook-controller:pr-449

jiridanek commented:
@daniellutz can you fix the odh-notebook-controller-e2e tests on ocp-ci?

![image](https://github.com/user-attachments/assets/cae1f75e-da2f-4a00-a442-df4e45987eca)

after that's done, I'll take another look. Wdyt?

daniellutz committed:
Add RBAC role to have access to OAuthClient objects

jiridanek commented:
Played with

```
    workbenches:
      devFlags:
        manifests:
          - contextDir: components/odh-notebook-controller/config
            sourcePath: ''
            uri: 'https://github.com/daniellutz/odh-kubeflow/archive/new-authorize-access-ui.tar.gz'
      managementState: Managed
```

for a while, and things seemed to work. I don't understand why these devFlags use image

```
quay.io/dlutz/odh-notebook-controller:new-authorize-access-ui
```

and not

```
quay.io/opendatahub/odh-notebook-controller:pr-449
```

Also the openshift-ci e2e tests on this PR are still failing.

Given my experience with the devFlags setup, I say

/lgtm

but the tests really need fixing before this can be merged, wdyt.

jiridanek commented:
After fresh install of RHOAI-2.18 nightly, with the devFlags from above, and upgrading to the new notebook-controller, I got three errors in the logs of the controller

{noformat}
{"level":"error","ts":"2025-01-31T09:25:05Z","msg":"Unable to retrieve secret","controller":"notebook","controllerGroup":"kubeflow.org","controllerKind":"Notebook","Notebook":{"name":"rocm-tensorflow","namespace":"jdanek"},"namespace":"jdanek","name":"rocm-tensorflow","reconcileID":"497b27ee-8331-4c0b-805e-5351d49cee7f","secret":"rocm-tensorflow","error":"Secret \"rocm-tensorflow-oauth-client\" not found","stacktrace":"github.com/opendatahub-io/kubeflow/components/odh-notebook-controller/controllers.(*OpenshiftNotebookReconciler).createOAuthClient\n\t/workspace/odh-notebook-controller/controllers/notebook_oauth.go:360\ngithub.com/opendatahub-io/kubeflow/components/odh-notebook-controller/controllers.(*OpenshiftNotebookReconciler).ReconcileOAuthClient\n\t/workspace/odh-notebook-controller/controllers/notebook_oauth.go:278\ngithub.com/opendatahub-io/kubeflow/components/odh-notebook-controller/controllers.(*OpenshiftNotebookReconciler).Reconcile\n\t/workspace/odh-notebook-controller/controllers/notebook_contr...
{"level":"error","ts":"2025-01-31T09:25:05Z","msg":"Unable to handle OAuthClient creation / update","controller":"notebook","controllerGroup":"kubeflow.org","controllerKind":"Notebook","Notebook":{"name":"rocm-tensorflow","namespace":"jdanek"},"namespace":"jdanek","name":"rocm-tensorflow","reconcileID":"497b27ee-8331-4c0b-805e-5351d49cee7f","error":"Secret \"rocm-tensorflow-oauth-client\" not found","stacktrace":"github.com/opendatahub-io/kubeflow/components/odh-notebook-controller/controllers.(*OpenshiftNotebookReconciler).ReconcileOAuthClient\n\t/workspace/odh-notebook-controller/controllers/notebook_oauth.go:280\ngithub.com/opendatahub-io/kubeflow/components/odh-notebook-controller/controllers.(*OpenshiftNotebookReconciler).Reconcile\n\t/workspace/odh-notebook-controller/controllers/notebook_controller.go:230\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:119\nsig...
{"level":"error","ts":"2025-01-31T09:25:05Z","msg":"Reconciler error","controller":"notebook","controllerGroup":"kubeflow.org","controllerKind":"Notebook","Notebook":{"name":"rocm-tensorflow","namespace":"jdanek"},"namespace":"jdanek","name":"rocm-tensorflow","reconcileID":"497b27ee-8331-4c0b-805e-5351d49cee7f","error":"Secret \"rocm-tensorflow-oauth-client\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:227"}
{noformat}

Full log 
[odh-notebook-controller-manager-7fffd544fc-qhf6f-manager.log](https://github.com/user-attachments/files/18615742/odh-notebook-controller-manager-7fffd544fc-qhf6f-manager.log)

jiridanek commented:
After I enabled culling, got the expected messages

```
{"level":"info","ts":"2025-01-31T09:27:51Z","logger":"controllers.Notebook","msg":"Update blocked, notebook pod template would be changed by the webhook","notebook":"rocm-pytorch","namespace":"jdanek","diff":"{v1.PodSpec}.Volumes[5->?]: {Name:oauth-client VolumeSource:{HostPath:nil EmptyDir:nil GCEPersistentDisk:nil AWSElasticBlockStore:nil GitRepo:nil Secret:&SecretVolumeSource{SecretName:rocm-pytorch-oauth-client,Items:[]KeyToPath{},DefaultMode:*420,Optional:nil,} NFS:nil ISCSI:nil Glusterfs:nil PersistentVolumeClaim:nil RBD:nil FlexVolume:nil Cinder:nil CephFS:nil Flocker:nil DownwardAPI:nil FC:nil AzureFile:nil ConfigMap:nil VsphereVolume:nil Quobyte:nil AzureDisk:nil PhotonPersistentDisk:nil Projected:nil PortworxVolume:nil ScaleIO:nil StorageOS:nil CSI:nil Ephemeral:nil}} != <invalid reflect.Value>"}
```

Full log [odh-notebook-controller-manager-7fffd544fc-qhf6f-manager (1).log](https://github.com/user-attachments/files/18615825/odh-notebook-controller-manager-7fffd544fc-qhf6f-manager.1.log)

jiridanek commented:
Logging out and back in from a workbench started with old config still worked.

atheo89 commented:
Hello there, I did a review based on the given DevFlags tested on a several RHOAI versions (2.16 and 2.18 nightly) 

Bellow are my testing points:
- Functionality wise, the given code works as expected. I am able to spin up a notebook, login, and then it directs me to the JL notebook without present me the user access info. :heavy_check_mark: 
- I can see the two secret volumes `oauth-client` and `oauth-config` mounted the notebook pod. Moreover, if i delete the any secret from those they recreated. :heavy_check_mark: 
-  I took a look as well doing a upgrade from 2.16 to 2.18 but didn't observed any error logs on the odh-notebook controller, however i noticed when i tried to create a new notebook after this upgrade the same issue that you saw here:
![image](https://github.com/user-attachments/assets/2149097d-981a-4a3a-b41d-611367be6d65)

- Also something else that noticed is (maybe is an issue from the odh-operetor) when i removed the DevFlags, the odh-controller continue to show as image the `daniellutz/odh-kubeflow/archive/new-authorize-access-ui` and i waited a lot and restart it many times.. 
- Something else that would be nice to have before approve this work is to fix the odh-e2e tests

overall /lgtm

jiridanek commented:
> Also something else that noticed is (maybe is an issue from the odh-operetor) when i removed the DevFlags, the odh-controller continued to show as image the daniellutz/odh-kubeflow/archive/new-authorize-access-ui and i waited a lot and restart it many times..

@mlassak can you confirm ^^^ is expected behavior? I saw that too, but I assumed it is, so I did not ask.

mlassak commented:
> @mlassak can you confirm ^^^ is expected behavior? I saw that too, but I assumed it is, so I did not ask.

Hmm, on a first impression, that does not sound right. I tried adding the dev flag as above, then removed them, and the image changed both times. I only tried it on fresh RHOAI 2.16 install though. I'll look into it further

atheo89 commented:
> > @mlassak can you confirm ^^^ is expected behavior? I saw that too, but I assumed it is, so I did not ask.
> 
> Hmm, on a first impression, that does not sound right. I tried adding the dev flag as above, then removed them, and the image changed both times. I only tried it on fresh RHOAI 2.16 install though. I'll look into it further

Hey @mlassak It happens again today using another DevFlag, (it is not related to this specific PR) probably the issue seem to be introduced on RHOAI 1.18. The workaround to fix it was to delete the DSC and  create a new one.

zdtsw commented:
devFlag is a once operation.
i put what you want in the operator, and it wont undo that if you remove devFlags.
the behavior you observe here matches the design.

if you wanna rollback to what the official release has, 
remove devFlag then restart operator
or another complicated way: update devFlag to point to the "official release manifests" 
or the last way: uninstall opreator and re-install it
all these can get the same result.

daniellutz committed:
Fix pointer.Int32Ptr deprecation on oauth-client test

daniellutz committed:
Add OAuthClient role to odh-notebook-controller's kubebuilder config

jiridanek commented:
/lgtm

I think it's crucial to get this in before we're doing the next RHOAI stable. So this should be added to 2.18 soonest.

daniellutz commented:
I was talking to @harshad16 yesterday and he thought it would be better to release in 2.19, without hurry.

daniellutz commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -264,3 +335,61 @@ func (r *OpenshiftNotebookReconciler) ReconcileOAuthRoute(
+				return fmt.Errorf("failed to get DataScienceCluster custom resource data: %w", err)
+			}
+			if err = r.Client.Patch(ctx, oauthClient, client.RawPatch(types.ApplyPatchType, data),
+				client.ForceOwnership, client.FieldOwner("rhods-operator")); err != nil {
```
I'm reviewing a couple of things regarding coverage, unit and e2e tests, I will check this as soon as possible as well. thanks for pointing out.

jiridanek replied:
can customers have rhods-operator-owned oauthclient? isn't this something that only appears when someone ran previous version of the pr where oauthclient was created by rhods-operator through the magic label?

atheo89 replied:
This is a leftover when this functionality was part of the `rhods-operator` https://github.com/opendatahub-io/opendatahub-operator/blob/84d22f35a620f43f0e8b397b1b45e2bdb25a8f46/controllers/secretgenerator/secretgenerator_controller.go#L250
imo now it should mention the odh-notebook-controller

harshad16 replied:
@daniellutz , can we fix this or can we remove the ownership bit here and utilize the above method?

daniellutz commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -264,3 +335,61 @@ func (r *OpenshiftNotebookReconciler) ReconcileOAuthRoute(
+		if apierrs.IsAlreadyExists(err) {
+			data, err := json.Marshal(oauthClient)
+			if err != nil {
+				return fmt.Errorf("failed to get DataScienceCluster custom resource data: %w", err)
```
I will change the wording, thanks!

daniellutz replied:
Changed. Thanks for the suggestion!

daniellutz committed:
Change error message on OAuth Client creation error

jiridanek commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -209,39 +227,92 @@ func NewNotebookOAuthSecret(notebook *nbv1.Notebook) *corev1.Secret {
+	for i := 0; i < SECRET_DEFAULT_COMPLEXITY; i++ {
+		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(letterRunes))))
+		if err != nil {
+			fmt.Printf("Error generating secret: %v\n", err)
```
this does not look right, if this errors, you cannot just log a message and continue, no?

daniellutz replied:
I will change this part right now. Thanks for the suggestion.

daniellutz committed:
Add validation when getting route from OAuthClient reconciler

daniellutz committed:
Fix pointer.Int32Ptr deprecation in webhook

jiridanek commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -264,3 +349,61 @@ func (r *OpenshiftNotebookReconciler) ReconcileOAuthRoute(
+			if err != nil {
+				return fmt.Errorf("failed to create OAuth Client: %w", err)
+			}
+			if err = r.Client.Patch(ctx, oauthClient, client.RawPatch(types.ApplyPatchType, data),
```
> QF1008: could remove embedded field "Client" from selector (staticcheck)

Not sure how I feel about this one. I'm of half mind disabling this lint check.

jiridanek commented:
/tetest

jiridanek commented:
/retest

jiridanek commented:
/retest ci/prow/odh-notebook-controller-e2e

jiridanek commented:
/test odh-notebook-controller-e2e

jiridanek commented:
/ok-to-test

atheo89 reviewed (commented):
Thanks for this work, Daniel!
Just one observation: when spinning up a notebook, it takes a bit of time before JupyterLab becomes available to the user. I noticed this behavior in two different notebooks—both required me to refresh the page 2–3 times before it was accessible.
It would be great if someone else could verify whether they’re experiencing the same behavior. Also, I didn’t notice any related errors or information in the logs regarding this latency.

jiridanek commented:
> Just one observation: when spinning up a notebook, it takes a bit of time before JupyterLab becomes available to the user. I noticed this behavior in two different notebooks—both required me to refresh the page 2–3 times before it was accessible.

This is known issue, sadly https://issues.redhat.com/browse/RHOAIENG-1576

jiridanek commented:
/lgtm

atheo89 commented:
> > Just one observation: when spinning up a notebook, it takes a bit of time before JupyterLab becomes available to the user. I noticed this behavior in two different notebooks—both required me to refresh the page 2–3 times before it was accessible.
> 
> This is known issue, sadly https://issues.redhat.com/browse/RHOAIENG-1576
> 
> In comments, I wrote at some point that this appears regularly, not just in perf tests under load.

Thank for pointing this but the issue says `we observed that ~10 times out of 100` not in every notebook spin up... 🤔

jiridanek commented:
I did observe every time in https://issues.redhat.com/browse/RHOAIENG-1576?focusedId=25042028&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-25042028

jiridanek commented:
Is this a regression for you? Could it be that since now going through the oauth dialogs is a little faster, there is greater likelihood to hit this?

atheo89 commented:
> Is this a regression for you? Could it be that since now going through the oauth dialogs is a little faster, there is greater likelihood to hit this?

Not sure, that's why wanted from other to take a look and see if that is normal for everyone. When i work on other stuff i don't see this often.

daniellutz committed:
Fix linter complaints about r.Client presence

daniellutz commented:
could this be server-related issue? like, powerful servers could be faster to providing the notebooks and less-powerful servers a delay could be observed?

maybe an issue to the dashboard team regarding when a notebook is "Ready" for real against "Ready" for the dashboard UI ? looks like they might be checking if it's ready in a wrong place then?

jiridanek commented:
Who knows. Let's merge, release, get a nightly, and play with it a bit more.

jiridanek commented:
Who knows. Let's merge, release, get a nightly, and play with it a bit more.

jiridanek commented:
/lgtm

harshad16 commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -264,3 +349,61 @@ func (r *OpenshiftNotebookReconciler) ReconcileOAuthRoute(
+		RedirectURIs: []string{"https://" + oauthClientRoute.Spec.Host},
+		GrantMethod:  oauthv1.GrantHandlerAuto,
+	}
+
```
lets add the ownerreference , same as we do everywhere else for consistency

```suggestion
         // Add .metatada.ownerReferences to the oauthclient to be deleted by
	// the Kubernetes garbage collector if the notebook is deleted
		if err := ctrl.SetControllerReference(notebook, oauthClient, r.Scheme); err != nil {
			log.Error(err, "Failed to set controller reference for oauthClient")
			return err
		}
```

harshad16 reviewed (changes requested):
We need to fix the ownership reference of rhods-operator

daniellutz committed:
Remove oauthClient ownership code regarding rhods-operator

daniellutz committed:
Add proper description to empty comments

daniellutz committed:
Add label for notebook-owner to the oauthclient object

harshad16 commented on components/odh-notebook-controller/controllers/notebook_oauth.go:
```diff
@@ -264,3 +349,63 @@ func (r *OpenshiftNotebookReconciler) ReconcileOAuthRoute(
+			if err != nil {
+				return fmt.Errorf("failed to create OAuth Client: %w", err)
+			}
+			if err = r.Patch(ctx, oauthClient, client.RawPatch(types.ApplyPatchType, data)); err != nil {
```
```suggestion
			if err = r.Patch(ctx, oauthClient, client.RawPatch(types.ApplyPatchType, data),
			client.ForceOwnership, client.FieldOwner("odh-notebook-controller")); err != nil {
```

client.ForceOwnership is needed, for applying the patch

harshad16 reviewed (changes requested):
On checking the patch scenario , i.e
- Create notebook and delete it and recreated 
without the following changes it fails with this error:
```
PatchOptions.meta.k8s.io \"\" is invalid: fieldManager: Required value: is required for apply patch
```

daniellutz committed:
Add ForceOwnership back to OAuthClient Patch

Co-authored-by: Harshad Reddy Nalla <harshadreddy16@gmail.com>

daniellutz committed:
Fix linter complaints about last commits

harshad16 reviewed (approved):
/lgtm
/approve

Tested the following scenarios;
- create a new notebook:
   - Application is accessible : :heavy_check_mark: 
   - oauthclient is created:  :heavy_check_mark:
- recreate a new notebook
   - Application is accessible : :heavy_check_mark: 
   - oauthclient is updated:  :heavy_check_mark:

harshad16 commented:
@daniellutz , please create a JIRA for OAuth client cleanup on the deletion of the notebook.
As that bit is not covered in this PR.
